### PR TITLE
[go_router] Add support for Iterable, List and Set to TypedGoRoute

### PR DESCRIPTION
**Pull request by Skogsfrae**
Add support for Iterable, List  and Set to TypedGoRoute

Changed signature of `GoRouteData.$location` for PR #2679
Resolves Issue [#108437](https://github.com/flutter/flutter/issues/108437)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@chunhtai From triage: ping on this review.

**Comment by chunhtai** [comment]
Updates: this one is waiting on https://github.com/flutter/packages/pull/2679

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
Hi @Skogsfrae looks like there are ci failures and merge conflict. can you rebase off the latest main and fix the ci?

**Comment by Skogsfrae** [comment]
Hi @chunhtai I finally managed to make the ci work. I temporarily modified the `analysis_options.yaml` to skip `go_router_builder` example app `.g.dart` files. I'll restore it in the `go_router_builder` pr after this gets merged

File: analysis_options.yaml
Comment (stuartmorgan) [inline analysis_options.yaml:21]: We should not be turning off analysis for an entire directory. What problem is this intended to solve?

Comment (stuartmorgan) [reply analysis_options.yaml:21]: > then I faced another problem for updating a file of a library without increasing its version, so I had to revert it.

The second failure you linked to is only for not having a CHANGELOG entry, not lack of version update, and as the error message says that's trivially overridable when it makes sense to do so (as in this case). Local ignores and an override for the changelog would be much better than changing the root analysis configuration to opt everything out.

**Comment by stuartmorgan** [comment]
Overriding for `go_router_builder`; ignores are dev-only changes.

**Comment by Skogsfrae** [comment]
@chunhtai this is ready to be merged

**Review by johnpryan — APPROVED** [review]

**Comment by stuartmorgan** [comment]
@johnpryan Who should do the second review on this?

**Comment by stuartmorgan** [comment]
@chunhtai It looks like you reviewed this before, could you re-review as the second reviewer?

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
@Skogsfrae can you rebase off latest main?

**Comment by Skogsfrae** [comment]
@chunhtai I rebased both this pr and the one on go_router_builder

**Comment by Skogsfrae** [comment]
@chunhtai I just rebased it again with the last release. Is there something else that needs to be done to merge this PR?

**Comment by Skogsfrae** [comment]
@chunhtai @stuartmorgan @johnpryan can this PR be merged so I can update the linked one and hopefully have that merged too?
I'm keeping this updated with upstream and I'd really like to be able to use this feature with go_router_builder

Thank you guys 🙏🏼

**Comment by stuartmorgan** [comment]
Based on the last comments from reviewers it looks like this is ready to land and just didn't get the label.

**Pull request merged**
